src/shell/oom.c

```diff
@@ -0,0 +1,245 @@
+    char *me;
+    unsigned long count;
+
+    shell_debug ("wakeup");
+
+    /* Consume an event from inotify file descriptor.
+     * Ignore event contents since there is only one watch registered.
+     */
+    if (read (oom->inotify_fd, &evbuf, sizeof (evbuf)) < 0) {
+        shell_log_error ("error reading from inotify fd");
+        return;
+    }
+    /* Read memory.events.
+     */
+    if (!(me = read_file (oom->memory_events_path))
+        || parse_memory_events (me, "oom_kill", &count) < 0) {
+        shell_log_error ("error reading %s", oom->memory_events_path);
+        goto out;
+    }
+    /* If any new oom events have been recorded, log them.
+     */
+    if (oom->oom_kill < count) {
+        shell_log_error ("Memory cgroup out of memory: killed %lu task%s.",
+                         count - oom->oom_kill,
+                         count - oom->oom_kill > 1 ? "s" : "");
+        oom->oom_kill = count;
+    }
+out:
+    free (me);
+}
+
+static void oom_destroy (struct shell_oom *oom)
+{
+    if (oom) {
+        int saved_errno = errno;
+        flux_watcher_destroy (oom->w);
+        if (oom->watch_fd >= 0) {
+            inotify_rm_watch (oom->inotify_fd, oom->watch_fd);
+        }
+        if (oom->inotify_fd >= 0)
+            close (oom->inotify_fd);
+        free (oom->memory_events_path);
+        free (oom);
+        errno = saved_errno;
+    }
+}
+
+static struct shell_oom *oom_create (flux_shell_t *shell, char *path)
+{
+    struct shell_oom *oom;
+
+    if (!(oom = calloc (1, sizeof (*oom))))
+        return NULL;
+    oom->inotify_fd = -1;
+    oom->watch_fd = -1;
+    oom->shell = shell;
+    oom->memory_events_path = path; // takes ownership
+    if ((oom->inotify_fd = inotify_init1 (IN_NONBLOCK | IN_CLOEXEC)) < 0
+        || (oom->watch_fd = inotify_add_watch (oom->inotify_fd,
+                                               oom->memory_events_path,
+                                               IN_MODIFY)) < 0)
+        goto error;
+    if (!(oom->w = flux_fd_watcher_create (shell->r,
+                                           oom->inotify_fd,
+                                           FLUX_POLLIN,
+                                           watch_cb,
+                                           oom)))
+        goto error;
+    flux_watcher_start (oom->w);
+    return oom;
+error:
+    oom_destroy (oom);
+    return NULL;
+}
+
+static int oom_init (flux_plugin_t *p,
+                      const char *topic,
+                      flux_plugin_arg_t *arg,
+                      void *data)
+{
+    flux_shell_t *shell = flux_plugin_get_shell (p);
+    struct shell_oom *oom;
+    char *path;
+
+    if (!(path = get_cgroup_path (getpid (), "memory.events", R_OK))) {
+        if (errno == ENOENT || errno == EPROTO )
+            return 0; // cgroup is not configured or not v2
+        return -1;
+    }
+    if (!shell || !(oom = oom_create (shell, path)))
+        return -1;
```

Comment:
I think if `!shell`, path won't be freed.  If `calloc` in `oom_create()` I don't think path won't be freed.

Reply:
Just pushed a fix for the mem leak.  This is new code so I sqaashed it (you were requesting a new commit?)

I could go either way on a new PR.  I  think @grondo was planning to test so I'll at least leave the shell plugin in place for the moment so he'll have that feedback during test (which I find pretty useful).

Reply:
Oh just realized you were speaking cave man.  "Ug find mem leak in new commit".  Never mind!

Reply:
will path be double free'd if oom_create() goes through the "goto error" path?

Reply:
Yes!  Sheesh, sorry.  Fixed and also a small bit of cosmetic cleanup.  I don't know why some of the generic shell plugin tests are now failing in CI but that might have to wait until tomorrow.

Reply:
This CI failure was reproducible on quartz, and by trial and error I found the error goes away if `oom` is loaded before `affinity`.  Now trying to figure out why...

Reply:
Ah it was a reliance on a residual errno value in the test.  Pushed a fix.